client/src/unifyfs.c

```diff
@@ -1113,6 +1122,7 @@ int unifyfs_gfid_read_reqs(read_req_t* in_reqs, int in_count)
      * we process it in batches as it comes in, eventually the
      * server will tell us it's sent us everything it can */
     int done = 0;
+    int rpc_done = 0;
     while (!done) {
         int tmp_rc = delegator_wait();
```

Comment:
The client can get stuck in here in a tight loop waiting for the server to reply with data here.  If the rpc causes a server error, will the client exit this?  I haven't checked the code thoroughly yet, but that was the worry before.

Reply:
Ugh, that time out we do have in delegator_wait is sort of meant to detect a dead server.  Looking at that, it can introduce another race condition that we should fix up.  Consider this situation.

The client issues a read rpc to server.  The server receives the rpc from the client and starts the read operation (successfully).  But... the server is slow to get results back to client and that 5 sec timeout expires.  The client assumes the server is dead and returns from ```delegator_wait``` with an error, which causes the client to exit this while loop early.  However, the server doesn't know that the client gave up waiting, and in the meantime, the server gets the results back from the read and it starts writing the read replies into the shared memory region.  The server might then deadlock waiting on the client if it fills the buffer, or the client might process those read replies for another read operation it issued later, which could lead to corrupt read results on the client side.

We should perhaps treat that "dead server" timeout to be a fatal error.  Or we need to fix up the hand-shaking between the client/server some here.

Reply:
There are probably many cases that we miss here in terms of handshaking between the server and client. Once the server starts fetching data, I do not think the server ever 'aborts/fails' a read request. And, like you said, the client cannot inform the server that 'I do not fetch data anymore.'  Basically, I do not think we have a clear way to handle exceptional cases gracefully between server and client. We need to fix that.

Also, I've tested reading random data back from clients, and the performance variance was huge. The following is the result from 8 compute nodes (each process executing 50 read operations of (random offset, 4MB)).

PPN | average | min | max | std
-- | -- | -- | -- | --
1 | 1.585 | 0.005 | 11.253 | 2.006
2 | 1.444 | 0.004 | 28.753 | 3.398
4 | 3.969 | 0.004 | 78.960 | 11.843

You can see a single 4MB read sometimes took more than 2mins when ppn was 4. I've also tested with ppn=8 (total 64 processes), and 55 processes couldn't even complete (50 times X 4MB read operations) in 30 minutes. I think we also need to rethink the server process model. It seems like a single delegator becomes a bottleneck when we have many processes per node, i.e., a large ppn.

I am bringing up this, because we may want to do something more than fixing the race between the server and client.